Cosmos relayer

denalimarsh opened the pull request:
This PR introduces a Cosmos relayer which subscribes to transactions/events on the Bridge blockchain through an rpc websocket and relays the appropriate transaction to the Bridge smart contracts. For example, a `MsgBurn` event on the Bridge will result in the validator sending a tx which contains relevant event data to CosmosBridge.sol's `NewProphecyClaim()` method. 

Note: follow the steps in the README in order to configure the environment required for cosmos relayer initialization.

Example initialization cmd:
`ebrelayer init cosmos [tendermint-node] [ethereum-provider] [bridge-registry-contract-address] [validator-private-key]`. 

Example relay of `MsgBurn` event:
<img width="665" alt="Screen Shot 2019-10-26 at 8 46 20 PM" src="https://user-images.githubusercontent.com/15370712/67629371-b8dba880-f831-11e9-9665-e976f44e0614.png">

This PR is blocked by https://github.com/cosmos/peggy/pull/82.

musnit commented:
@denalimarsh can you change the base of this PR to one which already has all commits unrelated to this PR, so that we only see relevant commits for the review? We can change base back to cosmos:master when the PR is unblocked by other PRs being merged there.

denalimarsh commented:
> @denalimarsh can you change the base of this PR to one which already has all commits unrelated to this PR, so that we only see relevant commits for the review? We can change base back to cosmos:master when the PR is unblocked by other PRs being merged there.

@musnit are there any available branches in cosmos/peggy which meet this criteria? This PR is based off musnit:burnn, which is on a different remote. Should we push a cosmos:burnn branch up in order to use it as the base merge branch for this PR?

musnit commented:
@denalimarsh have pushed that here temporarily: https://github.com/cosmos/peggy/tree/musnit/add_initial_burn_functionality

fedekunze commented on cmd/ebrelayer/main.go:
```diff
@@ -67,27 +70,43 @@ var rootCmd = &cobra.Command{
-		Short: "Initializes a web socket which streams live events from a smart contract",
+func initEthereumRelayerCmd() *cobra.Command {
+	initEthereumRelayerCmd := &cobra.Command{
+		Use:   "initEth [web3Provider] [contractAddress] [eventSignature] [validatorFromName] --chain-id [chain-id]",
```
you could maybe create an `init` Command and then use `AddCommand` to add two separate `ethereum` and `cosmos` commands. IMO it'd provide a better UX.

fedekunze replied:
```suggestion
		Use:   "ethereum [web3Provider] [contractAddress] [eventSignature] [validatorFromName] --chain-id [chain-id]",
```

denalimarsh replied:
Agreed, this UX is much better. Added the `init` subcommand in https://github.com/cosmos/peggy/pull/77/commits/eea99956698726a920b255fd281efa1eda41e914

fedekunze commented on cmd/ebrelayer/main.go:
```diff
@@ -67,27 +70,43 @@ var rootCmd = &cobra.Command{
+//
+func initCosmosRelayerCmd() *cobra.Command {
+	initCosmosRelayerCmd := &cobra.Command{
+		Use:     "initCos [tendermintProvider] [web3Provider] [contractAddress] [privateKey]",
```
```suggestion
		Use:     "cosmos [tendermintNode] [web3Provider] [contractAddress] [privateKey]",
```

Side question, are there any security implications with passing the private key directly?

denalimarsh replied:
Not directly since the cmd is run locally, but I think using an environment variable for the private key would be cleaner here.

denalimarsh replied:
https://github.com/cosmos/peggy/pull/77/commits/6f01e466c0910fb7574bfa0f8fe99cf47c944abb

fedekunze commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,79 @@
+			txRes := tx.Result
+			for i := 1; i < len(txRes.Events); i++ {
+				switch txRes.Events[i].Type {
+				case "burn":
```
we should use the event constants here

denalimarsh replied:
892aeaf655fcfde659b629fdc1ac826cb143a7fb, https://github.com/cosmos/peggy/pull/77/commits/61ba18be188ff2c2c174be0c53ac4b8fd0e385b9

fedekunze commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,79 @@
+	for {
+		select {
+		case result := <-out:
+			tx := result.Data.(tmtypes.EventDataTx)
```
can we check for errors on casting types?

denalimarsh replied:
https://github.com/cosmos/peggy/pull/77/commits/a3fccbe30ddc2c4edf6a31f30921d6b87e574eb8

fedekunze commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,79 @@
+		select {
+		case result := <-out:
+			tx := result.Data.(tmtypes.EventDataTx)
+			txIndex := result.Data.(tmtypes.EventDataTx).Index
```
ditto

denalimarsh replied:
`txIndex` has been removed in latest version

fedekunze commented on cmd/ebrelayer/txs/relayToEthereum.go:
```diff
@@ -0,0 +1,112 @@
+	auth := bind.NewKeyedTransactor(privateKey)
+	auth.Nonce = big.NewInt(int64(nonce))
+	auth.Value = big.NewInt(0)     // in wei
+	auth.GasLimit = uint64(300000) // 300,000 Gwei in units
```
gas limit should also be a const

denalimarsh replied:
Included in https://github.com/cosmos/peggy/pull/77/commits/61ba18be188ff2c2c174be0c53ac4b8fd0e385b9

fedekunze commented on cmd/ebrelayer/txs/relayToEthereum.go:
```diff
@@ -0,0 +1,112 @@
+	}
+
+	// Get the transaction receipt
+	receipt, err := client.TransactionReceipt(context.Background(), tx.Hash())
```
is there any way to export the receipt?

denalimarsh replied:
what do you mean by export the receipt? do you mean get more detailed tx data, such as emitted events?

musnit commented on .gitignore:
```diff
@@ -8,7 +8,9 @@
 # Build
 vendor
 .vendor-new
+node_modules/ 
```
duplicate

denalimarsh replied:
https://github.com/cosmos/peggy/pull/77/commits/c19d68a4eef90acac74b4b7b8aa13afe18f6fa90

musnit commented on README.md:
```diff
@@ -82,6 +83,9 @@ ebcli tx send validator $(ebcli keys show testuser -a) 10stake --chain-id=peggy
 ebcli query account $(ebcli keys show testuser -a) --trust-node
 
+# Then wait 10 seconds then confirm your validator was created correctly, and has become Bonded status
+ebcli query staking validators --trust-node
```
What's this doing here? This PR has a lot of weird commits in it too, seems like it may have gotten mixed up with other functionality?

denalimarsh replied:
I've changed the base branch to `base/bridge-contracts`, where I've merged the updated bridge contracts on top of `musnit/add-initial-burn-functionality`. It makes the commit history much cleaner and now it should be easier to review these changes.

musnit commented on README.md:
```diff
@@ -108,53 +112,287 @@ ebcli rest-server --trust-node
+yarn peggy:abi
+
+# Get contract's address
+yarn peggy:address
```
Same here, all these additions were already in the contract tooling branch?

denalimarsh replied:
`musnit/add-initial-burn-functionality` branch didn't have the contract tooling features yet, the new base branch does and this commit is no longer included in this PRs changes.

denalimarsh commented on app/app.go~HEAD:
```diff
@@ -0,0 +1,245 @@
+package app
```
https://github.com/cosmos/peggy/pull/77/commits/c19d68a4eef90acac74b4b7b8aa13afe18f6fa90

denalimarsh commented on cmd/ebrelayer/events/cosmosEvent.go:
```diff
@@ -0,0 +1,93 @@
+	// Iterate over each rune in the coin string
+	for i, char := range coinRune {
+		// Regex will match first letter [a-z] (lowercase)
+		matched, err := regexp.MatchString(`[a-z]`, string(char))
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/events/cosmosEvent.go:
```diff
@@ -0,0 +1,93 @@
+	// TODO: Add tokenContractAddress to MsgBurn event
+	tokenContractAddressString := "0xbeddb076fa4df04859098a9873591dce3e9c404d"
+	if !common.IsHexAddress(tokenContractAddressString) {
+		log.Fatal("Invalid token address: %v", tokenContractAddressString)
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/events/cosmosEvent.go:
```diff
@@ -0,0 +1,93 @@
+		// Regex will match first letter [a-z] (lowercase)
+		matched, err := regexp.MatchString(`[a-z]`, string(char))
+		if err != nil {
+			log.Fatal("Coin symbol/amount parsing error: %v", err)
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/events/cosmosEvent.go:
```diff
@@ -0,0 +1,93 @@
+
+	// Parse Ethereum receiver
+	if !common.IsHexAddress(eventData[1]) {
+		log.Fatal("Invalid recipient address: %v", eventData[1])
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,76 @@
+			for i := 1; i < len(txRes.Events); i++ {
+				event := txRes.Events[i]
+				eventName := string(txRes.Events[i].Type)
+				switch eventName {
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,76 @@
+					}
+				default:
+					// do nothing
+				}
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,76 @@
+		select {
+		case result := <-out:
+			tx := result.Data.(tmtypes.EventDataTx)
+			logger.Info("\t New transaction witnessed")
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,76 @@
+
+	query := "tm.event = 'Tx'"
+	out, err := client.Subscribe(context.Background(), "test", query, 1000)
+	if err != nil {
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,76 @@
+		logger.Error("Failed to start a client", "err", err)
+		os.Exit(1)
+	}
+	defer client.Stop()
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,76 @@
+	client := tmclient.NewHTTP(tendermintProvider, "/websocket")
+	client.SetLogger(logger)
+	err := client.Start()
+	if err != nil {
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/txs/relayToEthereum.go:
```diff
@@ -0,0 +1,97 @@
+	}
+
+	var txHash common.Hash
+	switch eventData.EventName {
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/txs/relayToEthereum.go:
```diff
@@ -0,0 +1,97 @@
+		if err != nil {
+			log.Fatal(err)
+		}
+		txHash = tx.Hash()
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/txs/relayToEthereum.go:
```diff
@@ -0,0 +1,97 @@
+
+	case "lock":
+		// TODO: Integrate with MsgLock feature
+	}
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/txs/relayToEthereum.go:
```diff
@@ -0,0 +1,97 @@
+	// Parse public key
+	publicKey := privateKey.Public()
+	publicKeyECDSA, ok := publicKey.(*ecdsa.PublicKey)
+	if !ok {
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/txs/relayToCosmos.go:
```diff
@@ -22,12 +22,12 @@ import (
 //		(cdc) of the Bridge blockchain are required.
 //
-func RelayEvent(chainID string, cdc *amino.Codec, validatorAddress sdk.ValAddress, moniker string, passphrase string, claim *types.EthBridgeClaim) error {
+func RelayToCosmos(chainID string, cdc *amino.Codec, validatorAddress sdk.ValAddress, moniker string, passphrase string, claim *types.EthBridgeClaim) error {
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/events/ethereumEvents.go:
```diff
@@ -22,21 +22,21 @@ func IsEventRecorded(txHash string) bool {
-// PrintEvents : prints all the claims made on this event
-func PrintEvents() {
+// PrintLockEvents : prints all the claims made on this event
+func PrintLockEvents() {
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/events/cosmosEvent.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+// NewMsgEvent : parses MsgEvent data
+func NewMsgEvent(eventName string, eventData [3]string) MsgEvent {
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/events/cosmosEvent.go:
```diff
@@ -0,0 +1,93 @@
+	if !common.IsHexAddress(eventData[1]) {
+		log.Fatal("Invalid recipient address: %v", eventData[1])
+	}
+	ethereumReceiver := common.HexToAddress(eventData[1])
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/events/cosmosEvent.go:
```diff
@@ -0,0 +1,93 @@
+	// Parse symbol, amount from coin
+	coinRune := []rune(eventData[2])
+	amount := new(big.Int)
+	var symbol string
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/events/cosmosEvent.go:
```diff
@@ -0,0 +1,93 @@
+		if matched {
+			amount, _ = amount.SetString(string(coinRune[0:i]), 10)
+			symbol = string(coinRune[i:len(coinRune)])
+			break
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/events/cosmosEvent.go:
```diff
@@ -0,0 +1,93 @@
+	if !common.IsHexAddress(tokenContractAddressString) {
+		log.Fatal("Invalid token address: %v", tokenContractAddressString)
+	}
+	tokenContractAddress := common.HexToAddress(tokenContractAddressString)
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented on cmd/ebrelayer/events/cosmosEvent.go:
```diff
@@ -0,0 +1,93 @@
+		// On first match, split the coin into (amount, symbol)
+		if matched {
+			amount, _ = amount.SetString(string(coinRune[0:i]), 10)
+			symbol = string(coinRune[i:len(coinRune)])
```
https://github.com/cosmos/peggy/pull/77/commits/97deb19c402f17fd28e22902df2450d2ea746ed1

denalimarsh commented:
Integrated latest version of bridge contracts from https://github.com/cosmos/peggy/pull/82 (excluding the new validator set from https://github.com/cosmos/peggy/pull/83) and added steps for cosmos relayer testing to the README.

musnit commented:
@denalimarsh to confirm, this is expected to be merged *after* #82 now, right?

denalimarsh commented:
> @denalimarsh to confirm, this is expected to be merged _after_ #82 now, right?

Yes this PR is blocked by https://github.com/cosmos/peggy/pull/82.

denalimarsh commented:
Added ProphecyClaim parsing into an OracleClaim, a process which includes the application of the validator's private key to generate the signature. OracleClaims are then sent to BridgeBank on the Ethereum network.

Since the two relayers are subscribed to different chains (ethereum, cosmos), I've added a flag ` --make-claims` to the ethereum relayer which takes a boolean value. By running the claim relayer in another terminal tab, it will automatically witness ProphecyClaims, sign them, and send an OracleClaim back to the contract.

Since making OracleClaim requires both relayers, there's a conversation to be had about refactoring them together so that users only ever have to run one `ebrelayer init` command. 

Full use of the functionality introduced in this PR is also blocked by https://github.com/cosmos/peggy/pull/86 (for different tokens in ProphecyClaims/OracleClaims)

fedekunze commented on cmd/ebrelayer/relayer/ethereum.go:
```diff
@@ -0,0 +1,122 @@
+	"fmt"
+	"log"
+
+	amino "github.com/tendermint/go-amino"
```
this should be "github.com/cosmos/cosmos-sdk/codec"

denalimarsh replied:
https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

fedekunze commented on cmd/ebrelayer/relayer/ethereum.go:
```diff
@@ -0,0 +1,122 @@
+	switch cosmosSupport {
+	case true:
+		targetContract = txs.CosmosBridge
+		eventName = "LogNewProphecyClaim"
```
use const for the events name

denalimarsh replied:
Updated this to `events.LogNewProphecyClaim.String()` as the string is already declared there

fedekunze commented on cmd/ebrelayer/relayer/ethereum.go:
```diff
@@ -0,0 +1,122 @@
+		eventSignature = contractABI.Events["LogNewProphecyClaim"].Id().String()
+	case false:
+		targetContract = txs.BridgeBank
+		eventName = "LogLock"
```
ditto

denalimarsh replied:
Updated this to `eventName = events.LogLock.String()`

fedekunze commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,97 @@
+				}
+
+				switch claimType {
+				case events.MsgBurn, events.MsgLock:
```
wrap the indented code below into a private function

denalimarsh replied:
Added function `handleBurnLockMsg(...)` in https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549. It uses new function `BurnLockEventToCosmosMsg()` which parses a Burn/Lock event's attribute values so that `NewCosmosMsg` is cleaner and doesn't contain any type casting.

denalimarsh replied:
Might make sense for you to review this section again and confirm everything is alright as I made a lot of changes to it in https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549. Thanks!

fedekunze commented on cmd/ebrelayer/events/types.go:
```diff
@@ -0,0 +1,22 @@
+
+const (
+	// MsgBurn : Cosmos event 'CosmosMsg' type MsgBurn
+	MsgBurn Event = iota
```
or you can use `Unsupported` as the first event so it will have the 0 assigned to it

denalimarsh replied:
Went with the second option https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

fedekunze commented on cmd/ebrelayer/main.go:
```diff
@@ -153,6 +189,46 @@ func RunRelayerCmd(cmd *cobra.Command, args []string) error {
+	contractAddress := common.HexToAddress(args[2])
+
+	// Initialize the relayer
+	err = relayer.InitCosmosRelayer(
```
you can `return` here

denalimarsh replied:
https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

fedekunze commented on cmd/ebrelayer/relayer/ethereum.go:
```diff
@@ -0,0 +1,122 @@
+	if err != nil {
+		return err
+	}
+	fmt.Printf("\nSubscribed to %v contract at address: %s\n", targetContract, targetAddress.Hex())
```
can we standardize the commands to use a logger or plain print statements? I prefer the former if you ask me

denalimarsh replied:
What would be the ideal logger solution? go-ethereum has https://github.com/ethereum/go-ethereum/tree/master/log, but right now `InitCosmosRelayer()` uses a tendermint logger from github.com/tendermint/tendermint/libs/log. I favor a unified logging solution so we don't duplicate work if there's a future PR to combine the cosmos/ethereum relayer (I've created an issue for this https://github.com/cosmos/peggy/issues/88).

denalimarsh replied:
Would it make sense for this to wait until the issue is addressed or should we try to implement something as part of this PR?

fedekunze commented on cmd/ebrelayer/txs/registry.go:
```diff
@@ -0,0 +1,88 @@
+
+const (
+	// Valset : valset contract
+	Valset ContractRegistry = iota
```
```suggestion
	Valset ContractRegistry = iota + 1
```

denalimarsh replied:
Added in https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549, along with updating `[d]` -> `[d-1]` in ContractRegistry's `String()` function

fedekunze commented on cmd/ebrelayer/txs/registry.go:
```diff
@@ -0,0 +1,88 @@
+)
+
+// ContractRegistry :
+type ContractRegistry int
```
you can use `byte` here instead
```suggestion
type ContractRegistry byte
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

fedekunze commented on cmd/ebrelayer/txs/registry.go:
```diff
@@ -0,0 +1,88 @@
+		log.Fatal(err)
+	}
+
+	switch target {
```
missing default + panic

denalimarsh replied:
https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

fedekunze commented on cmd/ebrelayer/txs/relayToCosmos.go:
```diff
@@ -22,13 +22,12 @@ import (
 //
-func RelayEvent(chainID string, cdc *amino.Codec, validatorAddress sdk.ValAddress, moniker string, passphrase string, claim *types.EthBridgeClaim) error {
-
+func RelayLockToCosmos(chainID string, cdc *amino.Codec, validatorAddress sdk.ValAddress, moniker string, passphrase string, claim *types.EthBridgeClaim) error {
```
split function params into multiple lines

fedekunze replied:
and use cosmos-sdk/codec instead of tendermint/go-amino

denalimarsh replied:
https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

fedekunze commented on cmd/ebrelayer/txs/relayToEthereum.go:
```diff
@@ -0,0 +1,153 @@
+	client, auth, target := initRelayConfig(provider, contractAddress, event)
+
+	// Initialize CosmosBridge instance
+	fmt.Println("\nFetching CosmosBridge contract...")
```
use logger

denalimarsh replied:
see discussion https://github.com/cosmos/peggy/pull/77/#discussion_r342373546

fedekunze commented on cmd/ebrelayer/txs/relayToEthereum.go:
```diff
@@ -0,0 +1,153 @@
+	client, auth, target := initRelayConfig(provider, contractAddress, event)
+
+	// Initialize Oracle instance
+	fmt.Println("\nFetching Oracle contract...")
```
ditto logger

denalimarsh replied:
see discussion https://github.com/cosmos/peggy/pull/77/#discussion_r342373546

fedekunze commented on cmd/ebrelayer/txs/relayToEthereum.go:
```diff
@@ -0,0 +1,153 @@
+
+	var targetContract ContractRegistry
+
+	switch event {
```
missing default + panic

denalimarsh replied:
https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

fedekunze commented on testnet-contracts/scripts/sendCheckProphecy.js:
```diff
@@ -0,0 +1,84 @@
+  const totalPower = Number(result[2]);
+
+  if (result) {
+    console.log("\n\tProphecy " + prophecyID + " status:");
```
you can use a single `console.log` with backticks `

denalimarsh replied:
In https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549. Also updated lines 75-77 for consistency.

fedekunze commented on testnet-contracts/scripts/sendCheckProphecy.js:
```diff
@@ -0,0 +1,84 @@
+    console.log("Reached threshold:\t", valid);
+    console.log("----------------------------------------");
+  } else {
+    console.log("Error: no result from transaction!");
```
```suggestion
    console.error("Error: no result from transaction!");
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

fedekunze commented on testnet-contracts/scripts/sendProcessProphecy.js:
```diff
@@ -0,0 +1,82 @@
+  // Get event logs
+  const event = logs.find(e => e.event === "LogProphecyProcessed");
+  if (event) {
+    console.log("\n\tProphecy " + event.args._prophecyID + " processed");
```
ditto

denalimarsh replied:
In https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549. Also updated lines 74-76. For better consistency, does it make more sense to only use backticks with multiple variables or to use backticks everywhere (including `console.log` statements without variables)?

fedekunze commented on testnet-contracts/scripts/sendProcessProphecy.js:
```diff
@@ -0,0 +1,82 @@
+    console.log("Weighted signed power:\t", event.args._weightedSignedPower);
+    console.log("-------------------------------------------");
+  } else {
+    console.log("Error: no result from transaction!");
```
```suggestion
    console.error("Error: no result from transaction!");
```

denalimarsh replied:
https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

fedekunze commented on x/ethbridge/client/cli/tx.go:
```diff
@@ -59,3 +59,34 @@ func GetCmdCreateEthBridgeClaim(cdc *codec.Codec) *cobra.Command {
+				return err
+			}
+
+			msg := types.NewMsgBurn(cosmosSender, ethereumReceiver, amount)
```
validate basic

denalimarsh replied:
Might make more sense for these changes to be in https://github.com/cosmos/peggy/pull/79,  as it will introduce conflicts such as the one resolved in https://github.com/cosmos/peggy/pull/77/commits/e34c38433c1979ae1a8aeb427e87680008370e6e. @musnit can you confirm?

fedekunze commented on x/ethbridge/client/rest/rest.go:
```diff
@@ -31,10 +31,18 @@ type createEthClaimReq struct {
+}
+
 // RegisterRoutes - Central function to define routes that get registered by the main application
 func RegisterRESTRoutes(cliCtx context.CLIContext, r *mux.Router, storeName string) {
```
```suggestion
func RegisterRESTRoutes(cliCtx context.CLIContext, r *mux.Router, queryRoute string) {
```

denalimarsh replied:
See above re https://github.com/cosmos/peggy/pull/79

fedekunze commented on x/ethbridge/client/cli/tx.go:
```diff
@@ -59,3 +59,34 @@ func GetCmdCreateEthBridgeClaim(cdc *codec.Codec) *cobra.Command {
+func GetCmdBurn(cdc *codec.Codec) *cobra.Command {
+	return &cobra.Command{
+		Use:   "burn [cosmos-sender-address] [ethereum-receiver-address] [amount]",
+		Short: "burn some coins!",
```
I think I added a comment somewhere to fix this description to be more detailed cc: @musnit

fedekunze commented on testnet-contracts/test/test_valset.js:
```diff
@@ -109,6 +109,7 @@ contract("Valset", function(accounts) {
     });
 
     it("should not validate signatures on a different hashed message", async function() {
+      // TODO: Prefix this signature
```
create issue

denalimarsh replied:
created https://github.com/cosmos/peggy/issues/87 which contains more information

denalimarsh commented on cmd/ebrelayer/events/types.go:
```diff
@@ -0,0 +1,22 @@
+
+// String : returns the event type as a string
+func (d Event) String() string {
+	return [...]string{"burn", "lock", "LogLock", "LogNewProphecyClaim", "unsupported"}[d]
```
Still needed as +1 was not added to iota (see above)?

denalimarsh commented on x/ethbridge/types/msg_burn.go:
```diff
@@ -0,0 +1,64 @@
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// MsgBurnEth defines a message for burning ethereum and triggering a related event
```
https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

denalimarsh commented on x/ethbridge/types/msg_burn.go:
```diff
@@ -0,0 +1,64 @@
+	Amount           sdk.Coins       `json:"amount" yaml:"amount"`
+}
+
+// NewMsgBurnEth is a constructor function for MsgBurnEth
```
https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

denalimarsh replied:
This gave some merge conflicts which I resolved. Might be better for changes to the ethbridge module to be in https://github.com/cosmos/peggy/pull/79 or https://github.com/cosmos/peggy/pull/81

denalimarsh commented on cmd/ebrelayer/events/cosmosEvent.go:
```diff
@@ -0,0 +1,96 @@
+	}
+
+	// Parse token contract address
+	// TODO: Add tokenContractAddress to MsgBurn event
```
Added `tokenContractAddress` support to this PR in https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549. Will automatically populate using `token_contract_address` field from Burn/Lock event's attributes in https://github.com/cosmos/peggy/pull/86. Removed the default address, so all tokens will be `address(0)` until the other PR is merged.

denalimarsh commented on cmd/ebrelayer/main.go:
```diff
@@ -48,16 +49,26 @@ func init() {
 		return initConfig(rootCmd)
 	}
 
+	initCmd.PersistentFlags().String("make-claims", "", "Make oracle claims everytime a prophecy claim is witnessed")
```
https://github.com/cosmos/peggy/pull/86

denalimarsh commented on cmd/ebrelayer/main.go:
```diff
@@ -67,33 +78,64 @@ var rootCmd = &cobra.Command{
+	// Parse make claims boolean
+	var makeClaims bool
+
+	makeClaimsString := viper.GetString("make-claims")
```
https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

denalimarsh commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,97 @@
+				var claimType events.Event
+
+				// Parse event type from event name
+				if string(event.Type) == "burn" {
```
Added new function `getOracleClaimType(eventType string)` in https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

denalimarsh commented on cmd/ebrelayer/relayer/ethereum.go:
```diff
@@ -0,0 +1,122 @@
+					eventName, vLog.TxHash.Hex(), vLog.BlockNumber)
+
+				switch eventName {
+				case events.LogLock.String():
```
Added function `handleLogLockEvent(...)` in https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

denalimarsh commented on cmd/ebrelayer/relayer/ethereum.go:
```diff
@@ -0,0 +1,122 @@
+					if err != nil {
+						return err
+					}
+				case events.LogNewProphecyClaim.String():
```
Added function `handleLogNewProphecyClaimEvent(...)` in https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

denalimarsh replied:
Added function `handleLogNewProphecyClaimEvent(...)` in https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549

denalimarsh commented on cmd/ebrelayer/relayer/ethereum.go:
```diff
@@ -0,0 +1,122 @@
+)
+
+// InitEthereumRelayer : Starts an event listener on a specific Ethereum network, contract, and event
+func InitEthereumRelayer(cdc *amino.Codec, chainID string, provider string, contractAddress common.Address, cosmosSupport bool, validatorName string, passphrase string, validatorAddress sdk.ValAddress) error {
```
Updated in https://github.com/cosmos/peggy/pull/77/commits/be4c43930db5f73c577d2a55782bf464405dd549 and also split other long func declarations in ebrelayer

fedekunze commented:
rebased to master @denalimarsh. Mind fixing the conflicts 🙏

denalimarsh commented:
I've resolved the merge conflicts and the cosmos relayer is working correctly.

However, this PR is still blocked by https://github.com/cosmos/peggy/pull/86, which has been merged into open PR https://github.com/cosmos/peggy/pull/81. Once https://github.com/cosmos/peggy/pull/81 is merged to master, I'll be able to use the `FlagEthereumChainID` contained within that PR to fix the currently broken ethereum relayer.

denalimarsh commented:
I see the new conflicts, was expecting these after the https://github.com/cosmos/peggy/pull/81 merge. Will resolve these later today ✅

denalimarsh commented:
@fedekunze resolved conflicts

fedekunze commented on cmd/ebrelayer/relayer/network.go:
```diff
@@ -30,15 +30,12 @@ func SetupWebsocketEthClient(ethURL string) (*ethclient.Client, error) {
-			"invalid websocket eth client URL: %v",
-			ethURL,
-		)
+		return nil, errors.New("invalid websocket eth client URL")
```
```suggestion
		return nil, fmt.Errorf("invalid websocket eth client URL: %v",	ethURL)
```

denalimarsh replied:
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

fedekunze commented on cmd/ebrelayer/relayer/network.go:
```diff
@@ -30,15 +30,12 @@ func SetupWebsocketEthClient(ethURL string) (*ethclient.Client, error) {
 	client, err := ethclient.Dial(ethURL)
 	if err != nil {
-		return nil, fmt.Errorf("error dialing websocket client: %v", err)
+		return nil, errors.New("error dialing websocket client")
```
we can wrap the error here 
```suggestion
		return nil, errors.Wrap(err, "error dialing websocket client")
```

denalimarsh replied:
Is the `errors.Wrap` syntax compatible with go 1.13? This [go 1.13-errors](https://blog.golang.org/go1.13-errors) article on the golang.org blog recommends wrapping errors with %w.

denalimarsh replied:
Wrapped with %w in https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

fedekunze commented on cmd/ebrelayer/txs/registry.go:
```diff
@@ -0,0 +1,88 @@
+		}
+		return cosmosBridgeAddress, nil
+	default:
+		panic("Invalid target contract address")
```
```suggestion
		panic("invalid target contract address")
```

denalimarsh replied:
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

fedekunze commented on cmd/ebrelayer/txs/relayToCosmos.go:
```diff
@@ -9,33 +9,35 @@ package txs
+	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/cosmos/cosmos-sdk/x/auth/client/utils"
-
+	authtxb "github.com/cosmos/cosmos-sdk/x/auth/types"
```
nit, this should be 
```suggestion
	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
```

denalimarsh replied:
good find https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

fedekunze commented on cmd/ebrelayer/txs/relayToEthereum.go:
```diff
@@ -0,0 +1,162 @@
+	case events.LogNewProphecyClaim:
+		targetContract = Oracle
+	default:
+		panic("Invalid target contract address")
```
```suggestion
		panic("invalid target contract address")
```

denalimarsh replied:
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

fedekunze commented on cmd/ebrelayer/relayer/ethereum.go:
```diff
@@ -0,0 +1,181 @@
+	}
+
+	// Initiate the relay
+	err = txs.RelayLockToCosmos(chainID, cdc, validatorAddress, validatorName, passphrase, &prophecyClaim, rpcURL)
```
return here

denalimarsh replied:
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

fedekunze commented on cmd/ebrelayer/main.go:
```diff
@@ -52,16 +58,26 @@ func init() {
 	err := executor.Execute()
 	if err != nil {
-		panic(err)
+		log.Fatal("Failed executing CLI command: %s, exiting...\n", err)
```
```suggestion
		log.Fatal("failed executing CLI command: %s, exiting...\n", err)
```

denalimarsh replied:
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/relayer/ethereum.go:
```diff
@@ -0,0 +1,181 @@
+	if err != nil {
+		return err
+	}
+
```
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/relayer/ethereum.go:
```diff
@@ -0,0 +1,181 @@
+		return err
+	}
+
+	return nil
```
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/relayer/ethereum.go:
```diff
@@ -0,0 +1,181 @@
+	err := txs.RelayOracleClaimToEthereum(provider, contractAddress, events.LogNewProphecyClaim, oracleClaim)
+	if err != nil {
+		return err
+	}
```
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/relayer/ethereum.go:
```diff
@@ -0,0 +1,181 @@
+	// Initiate the relay
+	err := txs.RelayOracleClaimToEthereum(provider, contractAddress, events.LogNewProphecyClaim, oracleClaim)
+	if err != nil {
+		return err
```
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/relayer/ethereum.go:
```diff
@@ -0,0 +1,181 @@
+
+	// Initiate the relay
+	err := txs.RelayOracleClaimToEthereum(provider, contractAddress, events.LogNewProphecyClaim, oracleClaim)
+	if err != nil {
```
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/relayer/ethereum.go:
```diff
@@ -0,0 +1,181 @@
+	oracleClaim := txs.ProphecyClaimToSignedOracleClaim(event)
+
+	// Initiate the relay
+	err := txs.RelayOracleClaimToEthereum(provider, contractAddress, events.LogNewProphecyClaim, oracleClaim)
```
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,120 @@
+		return err
+	}
+
+	return nil
```
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,120 @@
+	if err != nil {
+		return err
+	}
+
```
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,120 @@
+	err := txs.RelayProphecyClaimToEthereum(web3Provider, contractAddress, claimType, prophecyClaim)
+	if err != nil {
+		return err
+	}
```
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,120 @@
+	// Relay the CosmosMsg to the Ethereum network
+	err := txs.RelayProphecyClaimToEthereum(web3Provider, contractAddress, claimType, prophecyClaim)
+	if err != nil {
+		return err
```
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,120 @@
+	// TODO: Need some sort of delay on this so validators aren't all submitting at the same time
+	// Relay the CosmosMsg to the Ethereum network
+	err := txs.RelayProphecyClaimToEthereum(web3Provider, contractAddress, claimType, prophecyClaim)
+	if err != nil {
```
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,120 @@
+
+	// TODO: Need some sort of delay on this so validators aren't all submitting at the same time
+	// Relay the CosmosMsg to the Ethereum network
+	err := txs.RelayProphecyClaimToEthereum(web3Provider, contractAddress, claimType, prophecyClaim)
```
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/relayer/cosmos.go:
```diff
@@ -0,0 +1,120 @@
+					if err != nil {
+						return err
+					}
+				case events.Unsupported:
```
Deleted, but keeping the switch statement as it will be useful later for different types of Cosmos msgs https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/main.go:
```diff
@@ -71,56 +87,78 @@ var rootCmd = &cobra.Command{
 	// Parse the address of the deployed contract
 	if !common.IsHexAddress(args[1]) {
-		return fmt.Errorf("Invalid contract-address: %v", args[1])
+		return errors.New("Invalid [bridge-contract-address]")
```
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/main.go:
```diff
@@ -71,56 +87,78 @@ var rootCmd = &cobra.Command{
 	ethereumProvider := args[0]
 	if !relayer.IsWebsocketURL(ethereumProvider) {
-		return fmt.Errorf("Invalid web3-provider: %v", ethereumProvider)
+		return errors.New("Invalid [web3-provider]")
```
https://github.com/cosmos/peggy/commit/52fdedf6f0b4e4b1c7cf822ab3519dabba3afa17

denalimarsh commented on cmd/ebrelayer/txs/parser.go:
```diff
@@ -42,31 +43,141 @@ func ParsePayload(valAddr sdk.ValAddress, event *events.LockEvent) (ethbridgeTyp
+	var symbol string
+
+	// Set up regex
+	isLetter, err := regexp.Compile(`[a-z]`)
```
https://github.com/cosmos/peggy/pull/77/commits/f3f67ca447d312dcf52542e7f8ec1af146648402

denalimarsh commented on cmd/ebrelayer/txs/utils.go:
```diff
@@ -0,0 +1,156 @@
+}
+
+// verifySignature: utility function for signature verification
+func verifySignature(hash common.Hash, signature []byte) {
```
https://github.com/cosmos/peggy/pull/77/commits/f3f67ca447d312dcf52542e7f8ec1af146648402

denalimarsh commented on x/ethbridge/client/rest/rest.go:
```diff
@@ -50,7 +50,14 @@ type burnOrLockEthReq struct {
 }
 
-// RegisterRoutes - Central function to define routes that get registered by the main application
+type burnEthReq struct {
```
https://github.com/cosmos/peggy/pull/77/commits/f3f67ca447d312dcf52542e7f8ec1af146648402

denalimarsh commented:
Fixed the module tests except for one:
```
=== FAIL: x/ethbridge TestMintSuccess (0.06s)
panic: ERROR:
Codespace: sdk
Code: 10
Message: "insufficient account funds;  < 10ethereum"
```
@musnit is CircleCI running these tests against the dockerized Bridge introduced in https://github.com/cosmos/peggy/pull/71?

musnit commented:
@denalimarsh looks like circleCI is running against this branch, 77?

anyway, we'll do a proper qa/fix of tests after this, as i think a few other things may have been missed during all these merges, so let's try get this merged then do that.
